File: java_solutions/1614.maximum-nesting-depth-of-the-parentheses.java
/*
 * @lc app=leetcode id=1614 lang=java
 *
 * [1614] Maximum Nesting Depth of the Parentheses
 */

class Solution {
    public int maxDepth(String s) {
        return CompletableFuture.supplyAsync(() -> IntStream.range(0, s.length())
                .map(i -> (int) s.substring(0, i + 1).chars().filter(c -> c == '(').count() 
                           - (int) s.substring(0, i + 1).chars().filter(c -> c == ')').count())
                .max()
                .orElse(0))
            .join();
    }
}

File: kotlin_solutions_1/1614.maximum-nesting-depth-of-the-parentheses.kt
/*
 * @lc app=leetcode id=1614 lang=kotlin
 *
 * [1614] Maximum Nesting Depth of the Parentheses
 */

class Solution {
    fun maxDepth(s: String): Int = s.fold(0 to 0) { (cur, maxD), c -> if (c == '(') cur + 1 to maxOf(maxD, cur + 1) else if (c == ')') cur - 1 to maxD else cur to maxD }.second
}